### PR TITLE
[Merged by Bors] - Add an example to test small window sizes

### PR DESCRIPTION
**Pull request by DJMcNab**
# Objective

We keep getting issues where things break at small window sizes, e.g #3368 (caused by #3153), #3596 ('caused' by #3545)

## Solution

- Add a test that we can make small windows.

Currently, this fails on my machine with some quite scary vulkan errors: 
```
2022-01-08T22:55:13.770261Z ERROR wgpu_hal::vulkan::instance: VALIDATION [VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 (0x7cd0911d)]
        Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 ] Object 0: handle = 0x1adbd410a60, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x7cd0911d | vkCreateSwapchainKHR() called with imageExtent = (225,60), which is outside the bounds returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(): currentExtent = (225,56), minImageExtent = (225,56), maxImageExtent = (225,56). The Vulkan spec states: imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface (https://vulkan.lunarg.com/doc/view/1.2.198.1/windows/1.2-extensions/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01274)
2022-01-08T22:55:13.770808Z ERROR wgpu_hal::vulkan::instance:   objects: (type: DEVICE, hndl: 0x1adbd410a60, name: ?)
2022-01-08T22:55:13.787403Z ERROR wgpu_hal::vulkan::instance: VALIDATION [VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 (0x7cd0911d)]
        Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 ] Object 0: handle = 0x1adbd410a60, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x7cd0911d | vkCreateSwapchainKHR() called with imageExtent = (225,56), which is outside the bounds returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(): currentExtent = (225,52), minImageExtent = (225,52), maxImageExtent = (225,52). The Vulkan spec states: imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface (https://vulkan.lunarg.com/doc/view/1.2.198.1/windows/1.2-extensions/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01274)
```
etc.

This might be a new issue here, although I'm surprised it's vulkan giving this error; wgpu should stop it if this is illegal.

### CODE REVIEW

File: examples/README.md [hunk 1]
```diff
@@ -255,6 +255,7 @@ Example | File | Description
 `scale_factor_override` | [`window/scale_factor_override.rs`](./window/scale_factor_override.rs) | Illustrates how to customize the default window settings
 `transparent_window` | [`window/transparent_window.rs`](./window/transparent_window.rs) | Illustrates making the window transparent and hiding the window decoration
 `window_settings` | [`window/window_settings.rs`](./window/window_settings.rs) | Demonstrates customizing default window settings
+`expanding_window` | [`window/expanding_window.rs`](./window/expanding_window.rs) | Demonstrates changing window settings at runtime. Also used to validate the bevy can handle arbitrarily small windows
```
Comment (alice-i-cecile) [inline examples/README.md:258]:
```suggestion
`expanding_window` | [`window/expanding_window.rs`](./window/expanding_window.rs) | Demonstrates changing window settings at runtime. Also used to validate that `bevy` can handle arbitrarily small windows
```

**Review by alice-i-cecile — APPROVED** [review]
Yes please. One nit, but LGTM.

**Comment by DJMcNab** [comment]
The first time (I think) that I've ever hoped that CI would fail.

File: examples/window/expanding_window.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
+use bevy::prelude::*;
+
+fn main() {
+    App::new()
+        .insert_resource(WindowDescriptor {
+            width: 200.,
```
Comment (rparrett) [inline examples/window/expanding_window.rs:6]: Could we throw the `200.`s into a couple consts?

**Comment by alice-i-cecile** [comment]
Works for me on Windows 10, with an Nvidia GTX 1070 card. Sorry!

**Comment by rparrett** [comment]
With the window size as 400x400 now, I think the example needs to be configured to run longer in CI

**Comment by DJMcNab** [comment]
Wait why?

How would the example not running in CI allow it to work as a canary in CI for small window sizes breaking?

**Review by alice-i-cecile — APPROVED** [review]

**Comment by DJMcNab** [comment]
Blocked on https://github.com/bevyengine/bevy/issues/3596

File: examples/window/expanding_window.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    let primary = windows.get_primary_mut().unwrap();
+    let height = primary.height();
+    let width = primary.width();
+    match *phase {
+        Phase::ContractingY => {
+            if height <= 0.5 {
+                *phase = ContractingX;
+            }
+            primary.set_resolution(width, (height - 4.).max(0.0))
```
Comment (bjorn3) [inline examples/window/expanding_window.rs:49]:
```suggestion
            primary.set_resolution(width, (height - 4.).max(1.0))
```
maybe this here and below works? May also need to be a couple of pixels bigger to account for window decorations.

Comment (DJMcNab) [reply examples/window/expanding_window.rs]: New version should avoid this

**Comment by rparrett** [comment]
`minimize` spelling seems to be preferred in bevy and winit. (sorry)

File: .github/example-run/minimising.ron [hunk 4]
```diff
@@ -0,0 +1,5 @@
+(
+    // We have no promised about frame rate.
```
Comment (alice-i-cecile) [inline .github/example-run/minimising.ron:2]:
```suggestion
    // We have no promises about frame rate.
```

File: .github/example-run/minimising.ron [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
+    // TODO: Make this as small as is feasible
```
Comment (alice-i-cecile) [inline .github/example-run/minimising.ron:3]: Can we address this now?

File: examples/window/minimising.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
+use std::time::Duration;
```
Comment (alice-i-cecile) [inline examples/window/minimising.rs:1]: Module level comment please.

File: examples/README.md [hunk 7]
```diff
@@ -307,6 +307,8 @@ Example | File | Description
 `scale_factor_override` | [`window/scale_factor_override.rs`](./window/scale_factor_override.rs) | Illustrates how to customize the default window settings
 `transparent_window` | [`window/transparent_window.rs`](./window/transparent_window.rs) | Illustrates making the window transparent and hiding the window decoration
 `window_settings` | [`window/window_settings.rs`](./window/window_settings.rs) | Demonstrates customizing default window settings
+`resizing` | [`window/resizing.rs`](./window/resizing.rs) | Demonstrates changing window size at runtime. Also used to validate that `bevy` can handle arbitrarily small windows
+`minimising` | [`window/minimising.rs`](./window/minimising.rs) | Demonstrates minimising a window from within your game. Also used to validate that `bevy` can handle minimised windows
```
Comment (alice-i-cecile) [inline examples/README.md:311]:
```suggestion
`minimising` | [`window/minimising.rs`](./window/minimising.rs) | Demonstrates minimising a window from within your game.
```

Move removed note to module level comment.

File: examples/README.md [hunk 8]
```diff
@@ -307,6 +307,8 @@ Example | File | Description
 `scale_factor_override` | [`window/scale_factor_override.rs`](./window/scale_factor_override.rs) | Illustrates how to customize the default window settings
 `transparent_window` | [`window/transparent_window.rs`](./window/transparent_window.rs) | Illustrates making the window transparent and hiding the window decoration
 `window_settings` | [`window/window_settings.rs`](./window/window_settings.rs) | Demonstrates customizing default window settings
+`resizing` | [`window/resizing.rs`](./window/resizing.rs) | Demonstrates changing window size at runtime. Also used to validate that `bevy` can handle arbitrarily small windows
```
Comment (alice-i-cecile) [inline examples/README.md:310]:
```suggestion
`resizing` | [`window/resizing.rs`](./window/resizing.rs) | Demonstrates changing window size at runtime.
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Just a few nits :)

File: examples/README.md (hunk 7 above)
Comment (alice-i-cecile) [reply examples/README.md:311]: I'd probably like to see this moved to some sort of integration tests folder eventually, but we can address that later once headless windows are working.

File: examples/window/minimising.rs [hunk 9]
```diff
@@ -0,0 +1,75 @@
+//! A test to confirm that `bevy` allows minimising the window
+//! This is run in CI to ensure that this doens't regress again.
```
Comment (KDecay) [inline examples/window/minimising.rs:2]:
```suggestion
//! This is run in CI to ensure that this doesn't regress again.
```

**Comment by DJMcNab** [comment]
It's worth noting that this hasn't actually caught the current minimisation regression - that's fixed in https://github.com/bevyengine/bevy/pull/4527. 

To have detected that, we would have needed https://github.com/bevyengine/bevy/pull/4437.

Also @alice-i-cecile can you confirm whether you approve this now?

File: examples/window/minimising.rs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    });
+    // camera
+    commands.spawn_bundle(PerspectiveCameraBundle {
+        transform: Transform::from_xyz(-2.0, 2.5, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
+        ..default()
+    });
+}
+
+/// A simple 2d scene, taken from the `rect` example
+fn setup_rect(mut commands: Commands) {
```
Comment (KDecay) [inline examples/window/minimising.rs:65]:
```suggestion
fn setup_2d(mut commands: Commands) {
```

File: examples/window/minimising.rs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    // it is currently.
+    App::new()
+        .insert_resource(WindowDescriptor {
+            title: "Minimising".into(),
+            ..Default::default()
+        })
+        .add_plugins(DefaultPlugins)
+        .add_system(minimise_automatically)
+        .add_startup_system(setup_3d)
+        .add_startup_system(setup_rect)
```
Comment (KDecay) [inline examples/window/minimising.rs:16]: I think calling this `setup_2d` is clearer then calling it like the example it's taken from.

```suggestion
        .add_startup_system(setup_2d)
```

**Comment by mockersf** [comment]
@cart mentioned in https://github.com/bevyengine/bevy/pull/4370#issuecomment-1085142023 that examples shouldn't be the place for this kind of tests

File: examples/window/resizing.rs [hunk 12]
```diff
@@ -0,0 +1,150 @@
+//! A test to confirm that `bevy` allows setting the window to arbitrary small sizes
+//! This is run in CI to ensure that this doens't regress again.
```
Comment (KDecay) [inline examples/window/resizing.rs:2]:
```suggestion
//! This is run in CI to ensure that this doesn't regress again.
```

**Review by IceSentry — APPROVED** [review]
The minimizing example crashes for me on windows 10, but it also happens with other 2d examples, so CI should now be able to detect it.

Edit: I missed the discussion on where this should be. I agree that it should probably be in some kind of integration test and not an example.

File: examples/window/resizing.rs [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    });
+    // camera
+    commands.spawn_bundle(PerspectiveCameraBundle {
+        transform: Transform::from_xyz(-2.0, 2.5, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
+        ..default()
+    });
+}
+
+/// A simple 2d scene, taken from the `rect` example
+fn setup_rect(mut commands: Commands) {
```
Comment (KDecay) [inline examples/window/resizing.rs:140]:
```suggestion
fn setup_2d(mut commands: Commands) {
```

File: examples/window/resizing.rs [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    // TODO: Debug and fix this if feasible
+    if !*first_complete {
+        *first_complete = true;
+        return;
+    }
+    let height = windows.height;
+    let width = windows.width;
+    match *phase {
+        Phase::ContractingY => {
+            if windows.height <= 1 {
```
Comment (KDecay) [inline examples/window/resizing.rs:64]:
```suggestion
            if height <= MIN_HEIGHT {
```

File: examples/window/resizing.rs [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+                *phase = ContractingX;
+            } else {
+                windows.height -= 4;
+            }
+        }
+        Phase::ContractingX => {
+            if width <= 1 {
```
Comment (KDecay) [inline examples/window/resizing.rs:71]:
```suggestion
            if width <= MIN_WIDTH {
```

File: examples/window/resizing.rs [hunk 16]
```diff
@@ -0,0 +1,150 @@
+//! A test to confirm that `bevy` allows setting the window to arbitrary small sizes
+//! This is run in CI to ensure that this doens't regress again.
+
+use bevy::{input::system::exit_on_esc_system, prelude::*};
+
+// The smallest size reached is 1x1, as X11 doesn't support windows with a 0 dimension
+// TODO: Add a check for platforms other than X11 for 0xk and kx0, despite those currently unsupported on CI.
+const MAX_WIDTH: u16 = 401;
+const MAX_HEIGHT: u16 = 401;
+
```
Comment (KDecay) [inline examples/window/resizing.rs:10]:
```suggestion
const MIN_WIDTH: u16 = 1;
const MIN_HEIGHT: u16 = 1;
const RESIZE_STEP: u16 = 4;

```

File: examples/window/resizing.rs [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+        return;
+    }
+    let height = windows.height;
+    let width = windows.width;
+    match *phase {
+        Phase::ContractingY => {
+            if windows.height <= 1 {
+                *phase = ContractingX;
+            } else {
+                windows.height -= 4;
```
Comment (KDecay) [inline examples/window/resizing.rs:67]:
```suggestion
                windows.height -= RESIZE_STEP;
```

File: examples/window/resizing.rs [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+            }
+        }
+        Phase::ContractingX => {
+            if width <= 1 {
+                *phase = ExpandingY;
+            } else {
+                windows.width -= 4;
```
Comment (KDecay) [inline examples/window/resizing.rs:74]:
```suggestion
                windows.width -= RESIZE_STEP;
```

File: examples/window/resizing.rs [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+            }
+        }
+        Phase::ExpandingY => {
+            if height >= MAX_HEIGHT {
+                *phase = ExpandingX;
+            } else {
+                windows.height += 4;
```
Comment (KDecay) [inline examples/window/resizing.rs:81]:
```suggestion
                windows.height += RESIZE_STEP;
```

File: examples/window/resizing.rs [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+            }
+        }
+        Phase::ExpandingX => {
+            if width >= MAX_WIDTH {
+                *phase = ContractingY;
+            } else {
+                windows.width += 4;
```
Comment (KDecay) [inline examples/window/resizing.rs:88]:
```suggestion
                windows.width += RESIZE_STEP;
```

File: Cargo.toml [hunk 21]
```diff
@@ -615,6 +615,15 @@ path = "examples/window/transparent_window.rs"
 name = "window_settings"
 path = "examples/window/window_settings.rs"
 
+[[example]]
+name = "resizing"
+path = "tests/window/resizing.rs"
```
Comment (alice-i-cecile) [inline Cargo.toml:620]: This is so cursed, but kind of makes sense? @mockersf I'll defer to your judgement here.

**Comment by alice-i-cecile** [comment]
LGTM if @mockersf agrees on the strategy of defining examples that live in the tests directory.

File: Cargo.toml (hunk 21 above)
Comment (IceSentry) [reply Cargo.toml:620]: I don't get the point of doing that. Is it to make it easy to run manually?

Comment (DJMcNab) [reply Cargo.toml:620]: It's because we currently have a scripting setup to run examples. And so if these are examples, they can be run by that setup.

Comment (mockersf) [reply Cargo.toml:620]: I'm OK with it

**Review by mockersf — APPROVED** [review]

**Pull request closed without merging**